// @gatling.io/core/target/checks/jsonOfTypeMultipleFind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapCheckBuilderJsonOfTypeMultipleFind = void 0;
const multipleFind_1 = require("./multipleFind");
const wrapCheckBuilderJsonOfTypeMultipleFind = (_underlying) => ({
    ...(0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying),
    ofString: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofString()),
    ofBoolean: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofBoolean()),
    ofInt: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofInt()),
    ofLong: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofLong()),
    ofDouble: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofDouble()),
    ofList: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofList()),
    ofMap: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofMap()),
    ofObject: () => (0, multipleFind_1.wrapCheckBuilderMultipleFind)(_underlying.ofObject())
});
exports.wrapCheckBuilderJsonOfTypeMultipleFind = wrapCheckBuilderJsonOfTypeMultipleFind;
