// @gatling.io/core/target/checks/validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapCheckBuilderValidate = void 0;
const session_1 = require("../session");
const final_1 = require("./final");
const wrapCheckBuilderValidate = (_underlying) => ({
    ...(0, final_1.wrapCheckBuilderFinal)(_underlying),
    transform: (f) => (0, exports.wrapCheckBuilderValidate)(_underlying.transform(f)),
    transformWithSession: (f) => (0, exports.wrapCheckBuilderValidate)(_underlying.transformWithSession((0, session_1.underlyingXWithSessionTo)(f))),
    withDefault: (value) => (0, exports.wrapCheckBuilderValidate)(typeof value === "function"
        ? _underlying.withDefault((0, session_1.underlyingSessionTo)(value))
        : _underlying.withDefault(value)),
    withDefaultEL: (value) => (0, exports.wrapCheckBuilderValidate)(_underlying.withDefaultEl(value)),
    validate: (name, f) => (0, final_1.wrapCheckBuilderFinal)(_underlying.validate(name, (0, session_1.underlyingXWithSessionTo)(f))),
    is: (expected) => (0, final_1.wrapCheckBuilderFinal)(typeof expected === "function"
        ? _underlying.is((0, session_1.underlyingSessionTo)(expected))
        : _underlying.is(expected)),
    isEL: (expected) => (0, final_1.wrapCheckBuilderFinal)(_underlying.isEL(expected)),
    isNull: () => (0, final_1.wrapCheckBuilderFinal)(_underlying.isNull()),
    not: (expected) => (0, final_1.wrapCheckBuilderFinal)(typeof expected === "function"
        ? _underlying.not((0, session_1.underlyingSessionTo)(expected))
        : _underlying.not(expected)),
    notEL: (expected) => (0, final_1.wrapCheckBuilderFinal)(_underlying.notEL(expected)),
    notNull: () => (0, final_1.wrapCheckBuilderFinal)(_underlying.notNull()),
    in: (expected, ...rest) => (0, final_1.wrapCheckBuilderFinal)(typeof expected === "function"
        ? _underlying.in((0, session_1.underlyingSessionTo)(expected))
        : _underlying.in([expected, ...rest])),
    inEL: (expected) => (0, final_1.wrapCheckBuilderFinal)(_underlying.inEL(expected)),
    exists: () => (0, final_1.wrapCheckBuilderFinal)(_underlying.exists()),
    notExists: () => (0, final_1.wrapCheckBuilderFinal)(_underlying.notExists()),
    optional: () => (0, final_1.wrapCheckBuilderFinal)(_underlying.optional()),
    lt: (value) => (0, final_1.wrapCheckBuilderFinal)(typeof value === "function" ? _underlying.lt((0, session_1.underlyingSessionTo)(value)) : _underlying.lt(value)),
    ltEL: (value) => (0, final_1.wrapCheckBuilderFinal)(_underlying.ltEL(value)),
    lte: (value) => (0, final_1.wrapCheckBuilderFinal)(typeof value === "function" ? _underlying.lte((0, session_1.underlyingSessionTo)(value)) : _underlying.lte(value)),
    lteEL: (value) => (0, final_1.wrapCheckBuilderFinal)(_underlying.lteEL(value)),
    gt: (value) => (0, final_1.wrapCheckBuilderFinal)(typeof value === "function" ? _underlying.gt((0, session_1.underlyingSessionTo)(value)) : _underlying.gt(value)),
    gtEL: (value) => (0, final_1.wrapCheckBuilderFinal)(_underlying.gtEL(value)),
    gte: (value) => (0, final_1.wrapCheckBuilderFinal)(typeof value === "function" ? _underlying.gte((0, session_1.underlyingSessionTo)(value)) : _underlying.gte(value)),
    gteEL: (value) => (0, final_1.wrapCheckBuilderFinal)(_underlying.gteEL(value))
});
exports.wrapCheckBuilderValidate = wrapCheckBuilderValidate;
