// @gatling.io/core/target/population.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapPopulationBuilder = void 0;
const duration_1 = require("./utils/duration");
const pauses_1 = require("./structure/pauses");
const session_1 = require("./session");
const wrapPopulationBuilder = (_underlying) => ({
    _underlying,
    protocols: (...protocols) => (0, exports.wrapPopulationBuilder)(_underlying.protocols(protocols.map((p) => p._underlying))),
    andThen: (...children) => (0, exports.wrapPopulationBuilder)(_underlying.andThen(children.map((c) => c._underlying))),
    disablePauses: () => (0, exports.wrapPopulationBuilder)(_underlying.disablePauses()),
    constantPauses: () => (0, exports.wrapPopulationBuilder)(_underlying.constantPauses()),
    exponentialPauses: () => (0, exports.wrapPopulationBuilder)(_underlying.exponentialPauses()),
    customPauses: (f) => (0, exports.wrapPopulationBuilder)(_underlying.customPauses((0, session_1.underlyingSessionTo)(f))),
    uniformPauses: (plusOrMinus) => (0, exports.wrapPopulationBuilder)(_underlying.uniformPauses((0, duration_1.toJvmDuration)(plusOrMinus))),
    pauses: (pauseType) => (0, exports.wrapPopulationBuilder)(_underlying.pauses((0, pauses_1.toJvmPauseType)(pauseType))),
    throttle: (...throttleSteps) => (0, exports.wrapPopulationBuilder)(_underlying.throttle(throttleSteps.map((t) => t._underlying))),
    noShard: () => (0, exports.wrapPopulationBuilder)(_underlying.noShard())
});
exports.wrapPopulationBuilder = wrapPopulationBuilder;
